### DISCUSSION TITLE
How to Ensure One Polygon Always Renders on Top of Another in iOS?

### DISCUSSION THREAD

**Discussion by davlh** [react-native-maps/react-native-maps#5191] [Q&A]
Hi team,

I've encountered an issue with rendering polygons using the native iOS MapKit provider. Specifically, I have two polygons:

**Background Polygon:** This polygon serves as a backdrop and fully encompasses the foreground polygon.
**Foreground Polygon:** This polygon should always render on top of the background polygon.

**The Problem:**

When the app loads, both polygons render correctly, with the foreground polygon displayed over the background polygon. However, when I update the fillColor of the background polygon in response to a user event (like a tap), the background polygon redraws and covers the foreground polygon, causing it to disappear from view.

What I've Tried:

1. **Rendering Order:**

I've ensured that the foreground polygon is rendered after the background polygon in the JSX, hoping this would keep it on top.
```
// Background Polygon
<MapView.Polygon
  coordinates={backgroundCoordinates}
  fillColor={backgroundFillColor} // This color changes based on user interaction
  strokeColor="rgba(0,0,0,1)"
/>

// Foreground Polygon
<MapView.Polygon
  coordinates={foregroundCoordinates}
  fillColor="rgba(0, 255, 0, 0.5)"
  strokeColor="rgba(0,0,0,1)"
/>
```
Unfortunately, updating the fillColor of the background polygon still causes it to render over the foreground polygon.

2. **Using Holes in the Background Polygon:**

I attempted to create a hole in the background polygon that matches the shape and position of the foreground polygon.

```
// Background Polygon with hole
<MapView.Polygon
  coordinates={backgroundCoordinates}
  holes={[foregroundCoordinates]}
  fillColor={backgroundFillColor}
  strokeColor="rgba(0,0,0,1)"
/>

// Foreground Polygon
<MapView.Polygon
  coordinates={foregroundCoordinates}
  fillColor="rgba(0, 255, 0, 0.5)"
  strokeColor="rgba(0,0,0,1)"
/>
```

This approach didn't work as expected because when I apply a fillColor to the background polygon, the hole area gets filled in as well, making the foreground polygon invisible. I can see the hole when I just specify a strokeColor and not fillColor so I know the holes prop is doing something, but whenever I add a fillColor, the hole gets filled in.

3. **Using zIndex Property:**

I tried using the zIndex property to control the rendering order of the polygons.

```
// Background Polygon
<MapView.Polygon
  coordinates={backgroundCoordinates}
  fillColor={backgroundFillColor}
  strokeColor="rgba(0,0,0,1)"
  zIndex={1}
/>

// Foreground Polygon
<MapView.Polygon
  coordinates={foregroundCoordinates}
  fillColor="rgba(0, 255, 0, 0.5)"
  strokeColor="rgba(0,0,0,1)"
  zIndex={999}
/>
```
However, this didn't resolve the issue. From what I've read, the zIndex property might not be supported for polygons on iOS in react-native-maps.

**My Questions:**

- Is there a way to ensure that the foreground polygon always renders on top of the background polygon, even after the background polygon updates?
- Is there a proper method to use the holes property so that the hole remains transparent and doesn't get filled when a fillColor is applied?
- Are there any workarounds or alternative approaches to achieve the desired rendering order on iOS using the native MapKit provider?

**Additional Details:**

- I'm using an older version of react-native-maps, specifically: react-native-maps@0.30.2. Upgrading may be a large undertaking so am hoping to find solution that works with this version.
- The polygons are defined with simple coordinate arrays.
- I need the background polygon's color to change based on user interactions, which is triggering the re-rendering issue.
- The issue may be specific to the iOS MapKit provider, and I haven't tested it on Android or with other providers.

**Any advice or solutions would be greatly appreciated!**

Thank you!

**Comment by perryfaro** [comment]
Unfortunately, iOS does not have the ability to order items based on zIndex, as far as we have been able to determine for a project. I think you should look into reconsidering how to handle the foreground and background items. If there is a relationship between them, you might consider that if the background needs to be updated, you also redraw the foreground.

**Reply by davlh** [reply]
Thanks for the response, the weird thing is I've tried forcing the foreground to redraw every time the background is updated, but the background still overwrites the foreground.
